includes/class-wc-payment-gateway-wcpay.php

```diff
@@ -1375,28 +1375,52 @@ public function process_payment_for_order( $cart, $payment_information, $schedul
 	 */
 	public function get_payment_method_to_use_for_intent() {
 		if ( WC_Payments_Features::is_upe_deferred_intent_enabled() ) {
-			return $this->get_payment_methods_from_request()[0];
+			$request_payment_method = sanitize_text_field( wp_unslash( $_POST['payment_method'] ?? '' ) ); // phpcs:ignore WordPress.Security.NonceVerification
+			return $this->get_payment_methods_from_gateway_id( $request_payment_method )[0];
 		}
 	}
 
 	/**
-	 * Get the payment methods used in the request.
+	 * Get payment method types to attach to intention request.
 	 *
+	 * @param Payment_Information $payment_information Payment information object for transaction.
 	 * @return array List of payment methods.
 	 */
-	private function get_payment_methods_from_request() {
-		$upe_payment_method = sanitize_text_field( wp_unslash( $_POST['payment_method'] ?? '' ) ); // phpcs:ignore WordPress.Security.NonceVerification
+	public function get_payment_method_types( $payment_information ) {
+		$request_payment_method = sanitize_text_field( wp_unslash( $_POST['payment_method'] ?? '' ) ); // phpcs:ignore WordPress.Security.NonceVerification
+		$token                  = $payment_information->get_payment_token();
+
+		if ( ! empty( $request_payment_method ) ) {
+			$payment_methods = $this->get_payment_methods_from_gateway_id( $request_payment_method );
+		} elseif ( ! is_null( $token ) ) {
+			$order           = $payment_information->get_order();
+			$order_id        = is_a( $order, 'WC_Order' ) ? $order->get_id() : null;
+			$payment_methods = $this->get_payment_methods_from_gateway_id( $token->get_gateway_id(), $order_id );
+		} else {
+			$payment_methods = WC_Payments::get_gateway()->get_payment_method_ids_enabled_at_checkout( null, true );
+		}
+
+		return $payment_methods;
+	}
 
-		if ( ! empty( $upe_payment_method ) && 'woocommerce_payments' !== $upe_payment_method ) {
-			$payment_methods = [ str_replace( 'woocommerce_payments_', '', $upe_payment_method ) ];
+	/**
+	 * Get the payment methods used in the request.
+	 *
+	 * @param string $gateway_id ID of processing payment gateway.
+	 * @param int    $order_id ID of related order, if applicable.
+	 * @return array List of payment methods.
+	 */
+	public function get_payment_methods_from_gateway_id( $gateway_id, $order_id = null ) {
+		if ( 'woocommerce_payments' !== $gateway_id ) {
```

Comment:
I believe we should check here is the id start with `woocommerce_payments_` as well. WDYT?

Reply:
Also, I lack a comment here explaining which gateways have this ID. Can you add it please?

Reply:
At this point, we know that the ID will start with `woocommerce_payments`; if the ID is anything other than `woocommerce_payments`, we presume that it will begin with `woocommerce_payments_` as this should be the only remaining available option. 

I simplified this (with a few explanatory comments) in 178d0760d108485e4ffa4debf73c0fa62ad35969 by simply checking whether the ID begins with `woocommerce_payments_` first instead. However, I don't really feel a need to also confirm that the ID begins with `woocommerce_payments`, as I don't believe that it's possible for neither of those cases to be `true`.